tests/test_numpy_dtypes.cpp

```diff
@@ -70,6 +70,13 @@ struct StringStruct {
     std::array<char, 3> b;
 };
 
+struct ArrayStruct {
+    char a[3][4];
```

Comment:
Zero-sized built-in arrays are illegal in C++ (GCC and Clang accept them, but std::is_array is false for them, so they can't easily be supported).

std::array<T, 0> is legal (C++ says so explicitly) and I've started writing some extra code to support them, but I now think they should be disallowed because they're causing all kinds of problems with numpy. `_dtype_from_pep3118` is *non-deterministic* with zero-sized arrays (I'm guessing because offsets are not unique so do not sort consistently). Given `^T{f:d:(0)I:empty:4x}`, it sometimes gives
```
[('d', '<f4'), ('empty', '<u4', (1,)), ('', 'V4')]
```
and sometimes gives
```
{'names':['d','pad0',''], 'formats':['<f4','V4',('<u4', (0,))], 'offsets':[0,4,4], 'itemsize':8}
```
In the latter case, the dtype match sanity check fails, presumably because the padding is now named `pad0`.